flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java

```diff
@@ -0,0 +1,197 @@
+
+	@Override
+	public void grantLeadership(UUID leaderSessionID) {
+		runActionIfRunning(() -> startNewDispatcherLeaderProcess(leaderSessionID));
+	}
+
+	private void startNewDispatcherLeaderProcess(UUID leaderSessionID) {
+		stopDispatcherLeaderProcess();
+
+		dispatcherLeaderProcess = createNewDispatcherLeaderProcess(leaderSessionID);
+
+		final DispatcherLeaderProcess newDispatcherLeaderProcess = dispatcherLeaderProcess;
+		FutureUtils.assertNoException(
+			previousDispatcherLeaderProcessTerminationFuture.thenRun(newDispatcherLeaderProcess::start));
+	}
+
+	private void stopDispatcherLeaderProcess() {
+		final CompletableFuture<Void> terminationFuture = dispatcherLeaderProcess.closeAsync();
+		previousDispatcherLeaderProcessTerminationFuture = FutureUtils.completeAll(
+			Arrays.asList(
+				previousDispatcherLeaderProcessTerminationFuture,
+				terminationFuture));
+	}
+
+	private DispatcherLeaderProcess createNewDispatcherLeaderProcess(UUID leaderSessionID) {
+		LOG.debug("Create new {} with leader session id {}.", DispatcherLeaderProcess.class.getSimpleName(), leaderSessionID);
+
+		final DispatcherLeaderProcess newDispatcherLeaderProcess = dispatcherLeaderProcessFactory.create(leaderSessionID);
+
+		forwardShutDownFuture(newDispatcherLeaderProcess);
+		forwardConfirmLeaderSessionFuture(leaderSessionID, newDispatcherLeaderProcess);
+
+		return newDispatcherLeaderProcess;
+	}
+
+	private void forwardShutDownFuture(DispatcherLeaderProcess newDispatcherLeaderProcess) {
+		newDispatcherLeaderProcess.getShutDownFuture().whenComplete(
+			(applicationStatus, throwable) -> {
+				synchronized (lock) {
+					// ignore if no longer running or if leader processes is no longer valid
+					if (running && this.dispatcherLeaderProcess == newDispatcherLeaderProcess) {
+						if (throwable != null) {
+							shutDownFuture.completeExceptionally(throwable);
+						} else {
+							shutDownFuture.complete(applicationStatus);
+						}
+					}
+				}
+			});
+	}
+
+	private void forwardConfirmLeaderSessionFuture(UUID leaderSessionID, DispatcherLeaderProcess newDispatcherLeaderProcess) {
+		FutureUtils.assertNoException(
+			newDispatcherLeaderProcess.getLeaderAddressFuture().thenAccept(
+				leaderAddress -> {
+					if (leaderElectionService.hasLeadership(leaderSessionID)) {
+						leaderElectionService.confirmLeadership(leaderSessionID, leaderAddress);
+					}
+				}));
+	}
+
+	@Override
+	public void revokeLeadership() {
+		runActionIfRunning(this::stopDispatcherLeaderProcess);
+	}
+
+	private void runActionIfRunning(Runnable runnable) {
+		synchronized (lock) {
+			if (running) {
+				runnable.run();
+			} else {
+				LOG.debug("Ignoring action because {} has already been stopped.", getClass().getSimpleName());
+			}
+		}
+	}
+
+	@Override
+	public void handleError(Exception exception) {
+		fatalErrorHandler.onFatalError(
+			new FlinkException(
+				String.format("Exception during leader election of %s occurred.", getClass().getSimpleName()),
```

Comment:
Including `getClass().getSimpleName()` not needed but also doesn't hurt I guess.

Reply:
True. Will remove it.